src/tls.rs

```diff
@@ -352,14 +384,40 @@ impl Accept for TlsAcceptor {
         cx: &mut Context<'_>,
     ) -> Poll<Option<Result<Self::Conn, Self::Error>>> {
         let pin = self.get_mut();
-        match ready!(Pin::new(&mut pin.incoming).poll_accept(cx)) {
-            Some(Ok(sock)) => {
-                let session = ServerSession::new(&pin.config.clone());
-                // let tls = Arc::new($this.config);
-                return Poll::Ready(Some(Ok(TlsStream::new(sock, session))));
+        // fast path for not upgrading
+        if !pin.upgrade {
+            match ready!(Pin::new(&mut pin.incoming).poll_accept(cx)) {
+                Some(Ok(sock)) => {
+                    let session = ServerSession::new(&pin.config.clone());
+                    // let tls = Arc::new($this.config);
+                    Poll::Ready(Some(Ok(TlsStream::new(sock, session, true))))
+                }
+                Some(Err(e)) => Poll::Ready(Some(Err(e))),
+                None => Poll::Ready(None),
+            }
+        } else if let Some(sock) = pin.socket.as_mut() {
+            let mut first_byte = [0];
+            match ready!(sock.poll_peek(cx, &mut first_byte)) {
```

Comment:
This will keep peeking at the first byte until `Ready`, until then it will keep `Pending`.
It doesn't need a timeout or anything, I'm assuming this handled somewhere else?

Reply:
I moved this comment because it's location is outdated.